src/Std/Time/DateTime/LocalDateTime.lean

```diff
@@ -0,0 +1,260 @@
+Subtracts `Month.Offset` from a `LocalDateTime`, it clips the day to the last valid day of that month.
+-/
+@[inline]
+def subMonthsClip (dt : LocalDateTime) (months : Month.Offset) : LocalDateTime :=
+  { dt with date := dt.date.subMonthsClip months }
+
+/--
+Adds a `Month.Offset` to a `LocalDateTime`, rolling over excess days to the following month if needed.
+-/
+def addMonthsRollOver (dt : LocalDateTime) (months : Month.Offset) : LocalDateTime :=
+  { dt with date := dt.date.addMonthsRollOver months }
+
+/--
+Subtracts a `Month.Offset` from a `LocalDateTime`, adjusting the day to the last valid day of the
+resulting month.
+-/
+@[inline]
+def subMonthsRollOver (dt : LocalDateTime) (months : Month.Offset) : LocalDateTime :=
+  { dt with date := dt.date.subMonthsRollOver months }
+
+/--
+Adds a `Month.Offset` to a `LocalDateTime`, rolling over excess days to the following month if needed.
+-/
+@[inline]
+def addYearsRollOver (dt : LocalDateTime) (years : Year.Offset) : LocalDateTime :=
+  { dt with date := dt.date.addYearsRollOver years }
+
+/--
+Subtracts a `Month.Offset` from a `LocalDateTime`, rolling over excess days to the following month if
+needed.
+-/
+@[inline]
+def addYearsClip (dt : LocalDateTime) (years : Year.Offset) : LocalDateTime :=
+  { dt with date := dt.date.addYearsClip years }
+
+/--
+Subtracts a `Year.Offset` from a `LocalDateTime`, this function rolls over any excess days into the
+following month.
+-/
+@[inline]
+def subYearsRollOver (dt : LocalDateTime) (years : Year.Offset) : LocalDateTime :=
+  { dt with date := dt.date.subYearsRollOver years }
+
+/--
+Subtracts a `Year.Offset` from a `LocalDateTime`, adjusting the day to the last valid day of the
+resulting month.
+-/
+@[inline]
+def subYearsClip (dt : LocalDateTime) (years : Year.Offset) : LocalDateTime :=
+  { dt with date := dt.date.subYearsClip years }
+
+/--
+Getter for the `Year` inside of a `LocalDateTime`.
+-/
+@[inline]
+def year (dt : LocalDateTime) : Year.Offset :=
+  dt.date.year
+
+/--
+Getter for the `Month` inside of a `LocalDateTime`.
+-/
+@[inline]
+def month (dt : LocalDateTime) : Month.Ordinal :=
+  dt.date.month
+
+/--
+Getter for the `Day` inside of a `LocalDateTime`.
+-/
+@[inline]
+def day (dt : LocalDateTime) : Day.Ordinal :=
+  dt.date.day
+
+/--
+Getter for the `Hour` inside of a `LocalDateTime`.
+-/
+@[inline]
+def hour (dt : LocalDateTime) : Hour.Ordinal dt.time.hour.fst :=
+  dt.time.hour.snd
+
+/--
+Getter for the `Minute` inside of a `LocalDateTime`.
+-/
+@[inline]
+def minute (dt : LocalDateTime) : Minute.Ordinal :=
+  dt.time.minute
+
+/--
+Getter for the `Second` inside of a `LocalDateTime`.
+-/
+@[inline]
+def second (dt : LocalDateTime) : Second.Ordinal dt.time.second.fst :=
+  dt.time.second.snd
+
+/--
+Getter for the `Second` inside of a `LocalDateTime`.
+-/
+@[inline]
+def nanosecond (dt : LocalDateTime) : Nanosecond.Ordinal :=
+  dt.time.nano
+
+/--
+Get the current monotonic time.
+-/
+@[inline]
+def now : IO LocalDateTime :=
```

Comment:
I'm not sure if you're planning to change some of this as you work out the time zone database stuff, but I think that there are currently still some holes in the `LocalDateTime` story:

- `LocalDateTime.now` queries the system clock in UTC and the formatting methods like `toLongDateFormatString` format it accordingly. I think this is just wrong, if I run `run_meta do println! (← LocalDateTime.now).toLongDateFormatString` then I should see exactly the time that I see in my task bar and not the current time in UTC.
- To use the formatting routines, I need to provide a time zone to convert the `LocalDateTime` into a point on the time line before formatting it. This should not be necessary. `LocalDateTime` literally represents "something that a clock shows" and I want to be able to format this directly. Also I think that the current syntax is too complicated for users. If I want to format an ISO 8601 date (pretending for a second that the `SQLDate` format isn't already defined) then it seems to me that I have to do something like `(date-spec% "YYYY-MM-DD" : Format .any).format (DateTime.ofLocalDateTime n .UTC)`. I think it would be good to explore whether this can be shortened, ideally to `n.format "YYYY-MM-DD"` (potentially at a bit of runtime cost compared to the compile-time format string).
- I think there is some arithmetic missing here. There is `addDays`, but I think we also need `addHours` etc. I'm guessing that this is probably not a 100% valid thing to do because leap seconds (?) but I think that the use-case is just too common to not offer this.